v2/internal/coord/routing/explore.go

```diff
@@ -0,0 +1,490 @@
+//
+// maxCpl is the maximum CPL (common prefix length) that will be scheduled.
+// interval is the base time interval to leave between explorations of the same CPL.
+// multiplier is a factor that is applied to interval for CPLs lower than the maximum to increase the interval between
+// explorations for lower CPLs (which contain nodes that are more distant).
+// jitter is a factor that is used to increase the calculated interval for the next explore
+// operation by a small random amount. It must be between 0 and 0.05. When zero, no jitter is applied.
+//
+// The interval to the next explore is calculated using the following formula:
+//
+//	interval + (maxCpl - CPL) x interval x multiplier + interval * rand(jitter)
+//
+// For example, given an a max CPL of 14, an interval of 1 hour, a multiplier of 1 and 0 jitter the following
+// schedule will be created:
+//
+//	CPL 14 explored every hour
+//	CPL 13 explored every two hours
+//	CPL 12 explored every three hours
+//	...
+//	CPL 0 explored every 14 hours
+//
+// For example, given an a max CPL of 14, an interval of 1 hour, a multiplier of 1.5 and 0.01 jitter the following
+// schedule will be created:
+//
+//	CPL 14 explored every 1 hour + up to 6 minutes random jitter
+//	CPL 13 explored every 2.5 hours + up to 6 minutes random jitter
+//	CPL 12 explored every 4 hours + up to 6 minutes random jitter
+//	...
+//	CPL 0 explored every 22 hours + up to 6 minutes random jitter
+func NewDynamicExploreSchedule(maxCpl int, start time.Time, interval time.Duration, multiplier float64, jitter float64) (*DynamicExploreSchedule, error) {
+	if maxCpl < 1 {
+		return nil, fmt.Errorf("maximum cpl must be greater than zero")
+	}
+
+	if interval < 1 {
+		return nil, fmt.Errorf("interval must be greater than zero")
+	}
+
+	if multiplier < 1 {
+		return nil, fmt.Errorf("interval multiplier must be greater than or equal to one")
+	}
+
+	if jitter < 0 {
+		return nil, fmt.Errorf("interval jitter must not be negative")
+	}
+
+	if jitter > 0.05 {
+		return nil, fmt.Errorf("interval jitter must not be greater than 0.05")
+	}
+
+	s := &DynamicExploreSchedule{
+		maxCpl:     maxCpl,
+		interval:   interval,
+		multiplier: multiplier,
+		jitter:     jitter,
+		cpls:       new(exploreList),
+	}
+
+	// build the initial schedule
+	for cpl := maxCpl; cpl >= 0; cpl-- {
+		*s.cpls = append(*s.cpls, &exploreEntry{
+			Cpl: cpl,
+			Due: start.Add(s.cplInterval(cpl)),
+		})
+	}
+	heap.Init(s.cpls)
+
+	return s, nil
+}
+
+func (s *DynamicExploreSchedule) NextCpl(ts time.Time) (int, bool) {
+	// is an explore due yet?
+	next := (*s.cpls)[0]
+	if !next.Due.After(ts) {
+		// update its schedule
+
+		interval := float64(s.interval) + float64(s.interval)*float64(s.maxCpl-next.Cpl)*s.multiplier
+		interval *= 1 + rand.Float64()*s.jitter
+
+		next.Due = ts.Add(s.cplInterval(next.Cpl))
+		heap.Fix(s.cpls, 0) // update the heap
+		return next.Cpl, true
+	}
+
+	return -1, false
+}
```

Comment:
I think it would be more go-like to unnest the happy-path:

```go
func (s *DynamicExploreSchedule) NextCpl(ts time.Time) (int, bool) {
	// is an explore due yet?
	next := (*s.cpls)[0]
	if next.Due.After(ts) {
	  return -1, false
	}
	
	...
}
```

Reply:
Agree, will make that change

Reply:
done